src/vision/status_quo/Alan_tries_to_debug_a_hang.md

```diff
@@ -0,0 +1,137 @@
+    at /home/alan/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.4.0/src/runtime/basic_scheduler.rs:145
+#19 0x0000555555582094 in tokio::runtime::Runtime::block_on (self=0x7fffffffd2f8, future=...)
+    at /home/alan/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.4.0/src/runtime/mod.rs:450
+#20 0x000055555557c22f in inventory_service::main () at /home/alan/code/inventory_service/src/main.rs:4
+```
+
+</details>
+
+Puzzled, the only line Alan even recognizes is the `main` entry point function for the service.
+He knows that async tasks in Rust aren't run individually on their own threads which allows them to scale better and use fewer resources but surely there has to be a thread somewhere that's running his code?
+Alan doesn't completely understand how async works in Rust but he's seen the `Future::poll` method so he assumes that there is a thread which constantly polls tasks to see if they are ready to wake up.
+"Maybe I can find that thread and inspect its state?" he thinks and then consults the cheatsheet for the appropriate command to see the threads in the program.
+`info threads` seems promising so he tries that:
+
+<details><summary>(gdb) info threads</summary>
+
+```
+(gdb) info threads
+  Id   Target Id                                          Frame 
+* 1    Thread 0x7ffff7c3b5c0 (LWP 1048) "inventory_servi" 0x00007ffff7d5e58a in epoll_wait (epfd=3, events=0x555555711340, 
+    maxevents=1024, timeout=49152) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
+```
+
+</details>
+
+Alan is now even more confused: "Where are my tasks?" he thinks.
+After looking through the cheatsheet and StackOverflow, he discovers there isn't a way to see which async tasks are waiting to be woken up in the debugger.
+Taking a shot in the dark, Alan concludes that this thread must be thread which is polling his tasks since it is the only one in the program.
+He googles "epoll_wait rust async tasks" but the results aren't very helpful and inspecting the stack frame doesn't yield him any clues as to where his tasks are so this seems to be a dead end.
+
+After thinking a bit, Alan realizes that since the runtime must know what tasks are waiting to be woken up, perhaps he can have the service ask the async runtime for that list of tasks every 10 seconds and print them to stdout? 
+While crude, this would probably also help him diagnose the hang.
+Alan gets to work and opens the runtime docs to figure out how to get that list of tasks.
+After spending 30 minutes reading the docs, looking at StackOverflow questions and even posting on users.rust-lang.org, he discovers this simply isn't possible and he will have to add tracing to his application to figure out what's going on.
+
+Disgruntled, Alan begins the arduous, boring task of instrumenting the application in the hope that the logs will be able to help him.
+
+## 🤔 Frequently Asked Questions
+
+
+* **What are the morals of the story?**
```

Comment:
The style of the FAQ sections was changed in #118 / #119, using sub headings (###) instead of nested bullet lists.